src/Lecture2.hs

```diff
@@ -288,8 +393,30 @@ It also can be:
 
 x + 45 + y
 
-Write a function that takes and expression and performs "Constant
+Write a function that takes an expression and performs "Constant
 Folding" optimization on the given expression.
+
+jakalx: The idea here is to remove all (Lit x) items, combine them and then
+form a new expression out of the remaining exression, if any and our final
+literal.
 -}
 constantFolding :: Expr -> Expr
-constantFolding = error "TODO"
+constantFolding e = case pullOutConstant e of
+  (0, Just x) -> x
+  (n, Nothing) -> Lit n
+  (n, Just x) -> Add (Lit n) x
+
+pullOutConstant :: Expr -> (Int, Maybe Expr)
```

Comment:
You can simplify the implementation if this function if you return `(Int, [String])` instead. And you'll only need to combine variable names to the nested addition in the end